includes/data-events/class-modal-checkout.php

```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * Newspack Data Events Modal Checkout helper.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack\Data_Events;
+
+use Newspack\Data_Events;
+use WP_Error;
+
+/**
+ * Class to register the Modal_Checkout listeners.
+ */
+final class Modal_Checkout {
+
+	/**
+	 * The name of the action for form submissions
+	 */
+	const FORM_SUBMISSION = 'form_submission_received';
+
+	/**
+	 * The name of the action for form submissions
+	 */
+	const FORM_SUBMISSION_SUCCESS = 'form_submission_success';
+
+	/**
+	 * Initialize the class by registering the listeners.
+	 *
+	 * @return void
+	 */
+	public static function init() {
+		Data_Events::register_listener(
+			'newspack_blocks_checkout_button_modal',
+			'modal_checkout_interaction',
+			[ __CLASS__, 'checkout_button_purchase' ]
+		);
+
+		Data_Events::register_listener(
+			'newspack_blocks_donate_block_modal',
+			'modal_checkout_interaction',
+			[ __CLASS__, 'donate_button_purchase' ]
+		);
+
+		Data_Events::register_listener(
+			'woocommerce_checkout_order_created',
+			'modal_checkout_interaction',
+			[ __CLASS__, 'checkout_attempt' ]
+		);
+
+		Data_Events::register_listener(
+			'newspack_blocks_modal_continue',
+			'modal_checkout_interaction',
+			[ __CLASS__, 'modal_pagination' ]
+		);
+	}
+
+	/**
+	 * Returns whether a product is a one time purchase, or recurring and when.
+	 *
+	 * @param string $product_id Product's ID.
+	 */
+	public static function get_purchase_recurrence( $product_id ) {
```

Comment:
Nonblocking nitpick: I believe we refer to this as `frequency` pretty consistently within the blocks codebase. Maybe we can maintain this terminology?

Reply:
I opted to leave this one -- it's a good point, but both GA4 and a lot of our other code for GA4 in Newspack Plugin use "recurrence", so I opted to leave it GA4-centric 🙂